zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/adhocsubprocess/AdHocSubProcessActivityActivateProcessor.java

```diff
@@ -0,0 +1,258 @@
+            .toList();
+    if (!elementsNotInAdHocSubProcess.isEmpty()) {
+      writeRejectionError(
+          command,
+          RejectionType.INVALID_STATE,
+          String.format(
+              ERROR_MSG_ELEMENTS_NOT_FOUND,
+              command.getValue().getAdHocSubProcessInstanceKey(),
+              elementsNotInAdHocSubProcess));
+
+      return;
+    }
+
+    // activate the elements
+    for (final var elementValue : command.getValue().getElements()) {
+      final var elementToActivate =
+          adHocSubprocessDefinition.getElementById(elementValue.getElementId());
+      final var elementInstanceRecord = new ProcessInstanceRecord();
+      elementInstanceRecord.wrap(adHocSubprocessElementInstance.getValue());
+      elementInstanceRecord
+          .setFlowScopeKey(adHocSubprocessElementInstance.getKey())
+          .setElementId(elementToActivate.getId())
+          .setBpmnElementType(elementToActivate.getElementType())
+          .setBpmnEventType(elementToActivate.getEventType());
+
+      // recalculating the tree path for element to be activated
+      final long elementToActivateInstanceKey = keyGenerator.nextKey();
+      final var elementTreePath =
+          new ElementTreePathBuilder()
+              .withElementInstanceProvider(elementInstanceState::getInstance)
+              .withCallActivityIndexProvider(processState::getFlowElement)
+              .withElementInstanceKey(elementToActivateInstanceKey)
+              .withFlowScopeKey(adHocSubprocessElementInstance.getKey())
+              .withRecordValue(elementInstanceRecord)
+              .build();
+      elementInstanceRecord
+          .setElementInstancePath(elementTreePath.elementInstancePath())
+          .setProcessDefinitionPath(elementTreePath.processDefinitionPath())
+          .setCallingElementPath(elementTreePath.callingElementPath());
+
+      commandWriter.appendFollowUpCommand(
+          elementToActivateInstanceKey,
+          ProcessInstanceIntent.ACTIVATE_ELEMENT,
+          elementInstanceRecord);
+    }
+
+    stateWriter.appendFollowUpEvent(
+        command.getKey(), AdHocSubProcessActivityActivationIntent.ACTIVATED, command.getValue());
+
+    responseWriter.writeEventOnCommand(
+        command.getKey(),
+        AdHocSubProcessActivityActivationIntent.ACTIVATED,
+        command.getValue(),
+        command);
+  }
+
+  @Override
+  public ProcessingError tryHandleError(
+      final TypedRecord<AdHocSubProcessActivityActivationRecord> command, final Throwable error) {
+    return switch (error) {
+      case final IllegalArgumentException e -> {
+        writeRejectionError(command, RejectionType.INVALID_ARGUMENT, e.getMessage());
+        yield ProcessingError.EXPECTED_ERROR;
+      }
+      case final IllegalStateException e -> {
+        writeRejectionError(command, RejectionType.INVALID_STATE, e.getMessage());
+        yield ProcessingError.EXPECTED_ERROR;
+      }
```

Comment:
💭 Do we still need to handle these cases, given we don't throw custom exceptions but write the rejections directly?

Reply:
I'm guess it's not needed anymore, indeed for the reason that we're not throwing/handling named errors. The default behaviour of this method is that it returns a `ProcessingError.UNEXPECTED_ERROR`. Which, in our case, would be applicable because if an exception occurs it would be unexpected within the context of our invariants (which are handled by appending the rejection.)

But perhaps @berkaycanbc could give more information regarding this?

Reply:
Done in 610247332d8d.